ToolBox/Polygon/PolygonDataQueueHandler.cs

```diff
@@ -629,40 +630,67 @@ private IEnumerable<TradeBar> GetTradeBars(HistoryRequest request)
                     tickerPrefix = string.Empty;
                     break;
             }
+            
+            var resolutionTimeSpan = request.Resolution.ToTimeSpan();
+            var lastRequestedBarStartTime = request.EndTimeUtc.RoundDown(resolutionTimeSpan);
+            var start = request.StartTimeUtc.Date;
+            var end = lastRequestedBarStartTime;
 
-            var start = request.StartTimeUtc;
-            var end = request.EndTimeUtc;
+            // Perform a check of the number of bars requested, this must not exceed a static limit
+            var aggregatesCountPerResolution = GetArrgegatesCountPerResoltion(request.Resolution);
+            var dataRequestedCount = (end - start).Ticks
+                                     / resolutionTimeSpan.Ticks / aggregatesCountPerResolution;
 
-            using (var client = new WebClient())
+            if (dataRequestedCount > BaseAggregateBarsLimit)
             {
-                var url = $"{HistoryBaseUrl}/v2/aggs/ticker/{tickerPrefix}{request.Symbol.Value}/range/1/{historyTimespan}/{start.Date:yyyy-MM-dd}/{end.Date:yyyy-MM-dd}?apiKey={_apiKey}";
-
-                var response = client.DownloadString(url);
+                end = start + TimeSpan.FromTicks(resolutionTimeSpan.Ticks * BaseAggregateBarsLimit / aggregatesCountPerResolution);
+                end = end.Date;
+            }
 
-                var result = JsonConvert.DeserializeObject<AggregatesResponse>(response);
-                if (result.Results == null)
+            while (start < lastRequestedBarStartTime.Date)
```

Comment:
Nice! 👍 
Believe this limitation also applies to other history requests too, posting some below. Suggest we push this logic into some private simple class which keeps the state of the current time and we can advance it and we reuse it for all the required cases
https://polygon.io/docs/get_v1_historic_forex__from___to___date__anchor
https://polygon.io/docs/get_v2_aggs_ticker__forexTicker__range__multiplier___timespan___from___to__anchor

https://polygon.io/docs/get_v2_ticks_stocks_trades__ticker___date__anchor
https://polygon.io/docs/get_v2_ticks_stocks_nbbo__ticker___date__anchor

Reply:
We could even reuse this class in other brokerages like `CoinApi` which had a similar limitation and logic I believe?

Reply:
Hi @Martin-Molinero thank you , all sounds reasonable, it was problematic to find a time to fix all this, but I will do as soon as I can, probably next week. thanks,

Reply:
I am making this a draft because this is a work in progress, requires to complete the logic to downloads historic equity ticks and tests  and then extend it to the other instances - quotes, forex, cryto